README.md
<p align="center">
  <img src="https://assets.gitlab-static.net/uploads/-/system/project/avatar/10888432/base-style-config.png"/>
</p>

<h1 align="center">Base coding style check configuration</h1>

<p align="center">
  <b>A Set of Essential Configuration Files for Backend/Frontend/Build code style checking.</b>
</p>

__________________

[![license](https://img.shields.io/github/license/mashape/apistatus.svg)](/LICENSE.txt)

This project is licensed under the terms of the [MIT license](/LICENSE.txt).
__________________

## Quick Start

* [Java style checking - Quick Start](./java/README.md#quick-start)
* [JS style checking - Quick Start](./js/README.md#quick-start)

__________________

## Goals

The idea is to have a common and "single" source of code styling rules, which can be used across different projects:

* Provides a **similar set of rules** for Backend's, Frontend's and Build's code.
* Provides a "single" source of configuration files.

## Features

* Provides a set of configuration files for **[Checkstyle](http://checkstyle.sourceforge.net)**, **[PMD](https://pmd.github.io)**, **[ESlint](https://eslint.org)** and **[CodeNarc](http://codenarc.sourceforge.net)**, bundle in a JAR:
  * A [Common Checkstyle's set](/config/common/common-checks.xml), for "all" files (.java, .js, .css, .gradle, etc.).
  * A [Checkstyle's set](/config/java/java-checks.xml), for Java's code.
  * A [PMD's set](/config/java/java-rules.xml), for Java's code.
  * A [ESLint's set](/config/js/.eslintrc.json), for JS's code.
    * A [ESLint's set for Typescript](/config/js/.typescript-eslintrc.json), for JS's code.
  * A [Codenarc's set](/config/groovy/groovy-rules.groovy) for Build's Gradle code and Groovy's code.
  * All sets define a similar set of rules in order to get the homogeneity in code style between Backend's, Frontend's and Build's code.
* Adds some new custom rules to Checkstyle, PMD adn Codenarc.

## Code Style Checking

Code Style Checking tries to catch differences on coding style.

* The idea is to guaranteed the code looks similar beyond developers preference.
* It will allowed to force some best practices and standard of coding across team development.
* It is then recommend should be used before code is submitted to repository by developers or is approved.

There are different development tools for doing this.

This project use **[Checkstyle](http://checkstyle.sourceforge.net)**, **[PMD](https://pmd.github.io)**, **[ESlint](https://eslint.org)** and **[CodeNarc](http://codenarc.sourceforge.net)**.

> Checkstyle is easy to use and more "light", and PMD is easy for adding new "powerful" rules.

## Defined Code Style Checking

The Defined Sets of rules are in:

* [Java style checking](./java/README.md)
* [JS style checking](./js/README.md)

### Code Style Checking Common rules

#### Common rules [common-checks.xml](src/config/common/common-checks.xml)

Will check all the files with extension: **c, cpp, css, conf, flowconfig, gitignore, gradle, groovy, h, hpp, java, js, jsx, json, html, md, mjs, pri, pro, properties, qml, sh, sql, ts, tsx, txt, xml, yml**[1].

* No tab characters on files.
* Requires a new line break at the end of the file[2].

> [1] Can be extended at [common-checks.xml](src/config/common/common-checks.xml).
> [2] In order to be friendly with Code Versioning tools.

#### Common Backend's, Frontend's and Build's code rules

To highlights:

* Naming convention:
  * CamelCase.
  * All in uppercase for constants.
    * Non-consecutive underscore (`_`) for Backend.
* Fields/Variables/Parameters/Methods names must have a length between 3 and 23 [1].
  * With some exceptions for Fields/Names/Parameters: `id`, `k` & `K` (useful for loops), some prepositions (`at`, `by`, `in`, `of`, `on`, `to` & `up`).
  * For Java/Groovy methods allows `of`, specially useful for factories.
  * Java/Groovy Test Code [3]:
    * Test instance methods names should start with `should`.
    * Other instance methods valid names are: `afterAll`, `afterEach`, `beforeAll` or `beforeEach`. [5]
      * Groovy: other instance methods must have a length between 3 and 23.
      * Java: static methods must have a length between 3 and 23.
* Classes/Constants names must have a length between 3 and 32 [1].
  * Test Class names will not have length limit.
* Type parameters names must be in uppercase, have a length between 3 and 23, starting with a letter.
* Brackets style: Stroustrup.
* Indentation: "Tab" character = 2 spaces.
* Line length limit: 160 characters per line.
  * Line with a long string are ignored.
* Method length limit: 30 lines per method [2].
  * Java/Groovy Code [3] will have a limit of 25 [4] public/protected/private methods (being a reasonable limit for functional implementations, which tend to have several small functions/methods).[5]
* Method parameters number limit: 7 parameters per method [6].
* Java Code [7] will have a limit of 7 fields by class.
* Class/File lines limit: 300 lines per class/file.
* Class count limit: 1 Class per file.
* Public fields must be `final`.
* Imports should be order alphabetically.
* Cyclomatic Complexity limit: 8 paths.
* Some spacing rules are the same.
* When wrapping operators, the operator must be at the beginning of a new line.

Combination of Method length limit and Class/File lines limit allows to:

* Have at most around of 10 big methods: ~10 methods of max ~30 lines or ~25 methods of max ~12 lines.
* Be functional, having small simple methods with single responsibility, and only a few "complex" methods of max 30 lines.
  * Several small private methods and few public methods.

> Some of these rules are ignored for Test files in order to give some flexibility, but its encourage in Source files, may also help to reduce the complexity of Test files.  
> [1] This 3/23/32 rule, allows for concatenated words, still readable, and it is easy to remember.  
> [2] Having 30 lines allows for seen all the code of a function/method in same screen page.  
> [3] ESLint does not yet provide a way of doing this.  
> [4] Although there exists public APIs that more than this amount of methods, e.g. `java.util.stream.Stream`.  
> [5] Although this rule can be circumvented using static function fields.  
> [6] Exception may be required in constructor , when building of complex inheritance with more than 7 fields as total, for special cases, e.g. using some bad designed but required API, in this case use a @SuppressWarnings("checkstyle:ParameterNumber"). In app own code, Composition should be used over Inheritance.  
> [7] CodeNarc and ESLint does not yet provide a way of doing this.

##### Semicolon

Semicolon is required by Java[1], but no by Groovy, then when having JS code, the use of the semicolon could be enforced based on the Backend language, in order to get some consistency between Backend and Frontend.

To enforce Java style, set the respective project's eslint configuration (`package.json` or `.eslintrc.*`) to:

```json
  "rules": {
    "semi": ["error", "always"]
  }
```

To enforce Groovy style, set the respective project's eslint configuration (`package.json` or `.eslintrc.*`) to:

```json
  "rules": {
    "semi": ["error", "never"]
  }
```

Although this rule is not set, the project should always set `semi` rule in order to guarantee consistency in the code:

```json
  "rules": {
    "semi": ["error", ".."]
  }
```

> [1] Although new programming languages don't require semicolon, may be forcing JS to use of semicolon can be a waste of typing.  

##### Quotes

The use of single `'` or double `"` quote string could be enforced based on the Backend language, in order to get some consistency between Backend and Frontend.

To enforce Java style, set the respective project's eslint configuration (`package.json` or `.eslintrc.*`) to:

```json
  "rules": {
    "quotes":  ["error", "double"]
  }
```

To enforce Groovy style, set the respective project's eslint configuration (`package.json` or `.eslintrc.*`) to:

```json
  "rules": {
    "quotes":  ["error", "single"]
  }
```

Although this rule is not set, the project should always set `quotes` rule in order to guarantee consistency in the code:

```json
  "rules": {
    "quotes":  ["error", ".."]
  }
```

#### Code Style convention/tips

When adding indentation for parameters of a constructor/method/function add an extra 2 spaces, which will differentiate it from the inner code:

e.g.:

Given:

```java
  public SomeConstructor(final SomeType parameter1, final @SomeAnnotation SomeTypeN parameterN) {
    this.field1 = parameter1;
    this.fieldN = parameterN;
  }
```

Make it:

```java
  public SomeConstructor(
      final SomeType parameter1,
      final @SomeAnnotation SomeTypeN parameterN) {
    this.field1 = parameter1;
    this.fieldN = parameterN;
  }
```

instead of:

```java
  public SomeConstructor(
    final SomeType parameter1,
    final @SomeAnnotation SomeTypeN parameterN) {
    this.field1 = parameter1;
    this.fieldN = parameterN;
  }
```
__________________

## Contributing

* **Use it**.
* **Share it**.
* **Fork it**.
* [Give it a Star](https://github.com/gmullerb/base-style-config).
* [Propose changes or improvements](https://github.com/gmullerb/base-style-config/issues).
* [Report bugs](https://github.com/gmullerb/base-style-config/issues).

## Evolution

[`CHANGELOG`](CHANGELOG.html): contains the information about changes in each version, chronologically ordered ([Keep a Changelog](http://keepachangelog.com)).

## License

[MIT License](/LICENSE.txt)
__________________

## Remember

* Use code style verification tools => Encourages Best Practices, Efficiency, Readability and Learnability.
* Code Review everything => Encourages Functional suitability, Performance Efficiency and Teamwork.
* If viable, Start testing early => Encourages Reliability and Maintainability.

## Additional words

Don't forget:

* **Love what you do**.
* **Learn everyday**.
* **Learn yourself**.
* **Share your knowledge**.
* **Think different!**.
* **Learn from the past, dream on the future, live and enjoy the present to the max!**.
* **Enjoy and Value the Quest** (It's where you learn and grow).

At life:

* Let's act, not complain.
* Be flexible.

At work:

* Let's give solutions, not questions.
* Aim to simplicity not intellectualism.


CHANGELOG.md
# Base Style Config Change Log

## October 2022

* Fixes CI configuration.

## May 2020 / base-style-config.jar - 4.1.1 / eslint-plugin-base-style-config - 2.1.1

* Updates Type parameters names regular expression.
* Updates README files.

## May 2020 / base-style-config.jar - 4.1.0 / eslint-plugin-base-style-config - 2.0.0

* Adds some prepositions to regular expressions for names: `at`, `by`, `in`, `on`, `to` & `up`.
* Updates README files.

## February 2020 / base-style-config.jar - 4.0.1 / eslint-plugin-base-style-config - 1.0.2

* Updates the Type Parameters names rule with an additional pattern to allow to given more context in the name.
* Updates README files.

## December 2019

* [eslint-plugin-base-style-config changes](js/CHANGELOG.md#1.0.1---December-2019)
* Updates README files.

## November 2019 / base-style-config.jar - 4.0.0 / eslint-plugin-base-style-config - 1.0.0

* Moves `js` style configuration files to its own project, it was hard to use this configuration files in npm projects (no gradle projects).
  * `js` and `java` project will have from now on each one its own version.
* typescript-eslint now extends from eslint rules
  * Improves maintainability.
* Improves eslint rule `id-match`, allowing `$` character at the beginning.
* Removes `@typescript-eslint/indent` was colliding with eslint `indent`.
* Removes `@typescript-eslint/no-angle-bracket-type-assertion` (It's no available anymore since 2.0.0).
  * Adds `@typescript-eslint/consistent-type-assertions`.
* Removes eslint rule `yoda`, same as `quotes` & `semicolon` should be enforce based on language and team, because it can also harm maintainability.
* Renames `coding-rules.xml` to `java-rules.xml`.
* Renames `coding-checks.xml` to `java-checks.xml`.
* Updates README file.

## 3.0.3 - September 2019

* Adds rules: `UseOnlyWillFamilyMethodsWhenStubbing`, `UseOnlyThenFamilyMethodsWhenMocking`, `DoNotMixBDDandTDD` and `PreferBDDTesting`.
* Renames `UseOnlyDoFamilyMethodsWhenMocking` rule to `UseDoFamilyMethodsWhenStubbing`.
* Fixes `LineLength` Codenarc rule, it was not ignoring single quoted strings.
* Updates README file.

## 3.0.2 - September 2019

* Standardize naming convention for Java, Groovy and JS:
  * Updates Codenarc rules: `ClassName`, `FieldName`, `MethodName`, `InterfaceName`, `ParameterName` , `PropertyName` and `VariableName`.
  * Updates Checkstyle rules: `LocalFinalVariableName`, `MemberName`, `MethodName`, `ParameterName` and `TypeName`.
  * Adds Checkstyle rules: `LambdaParameterName`, `ConstantName`.
  * Adds Custom Checkstyle rules: `ConstantNameLength` and `TestTypeName`.
  * Adds Custom Codenarc rules: `ConstantNameLength`, `GradleVariableName`, `GroovyTestClassName` and `GroovyTestMethodName`.
  * Adds `MethodNamingConventions` PMD rule for checking Test files.
  * Adds `id-length` ESlint rule.
  * Adds `@typescript-eslint/generic-type-naming` rule.
  * Updates Checkstyle rules: `ClassTypeParameterName`, `InterfaceTypeParameterName` and `MethodTypeParameterName`.
  * Updates `id-match` ESlint rule.
  * Excludes PMD rules: `FieldNamingConventions`, `FormalParameterNamingConventions`, `GenericsNaming`, `LocalVariableNamingConventions`, `LongVariable`, `ShortClassName`, `ShortVariable` and `ShortMethodName`, cases which are now resolved by Checkstyle rules.
  * Adds required suppressions.
* Updates `LineLength` Codenarc rule, to be synchronize with Checkstyle `LineLength` rule.

## 3.0.1 - August 2019

* Updates `TooManyMethods` PMD rule and `MethodCount` Codenarc rule to have a limit of 30 methods by class.
* Updates `ParameterNumber` Checkstyle rule and `TooManyFields` PMD rule to have be consistent with each other, having a limit of 7 fields by class and 7 parameters by method (including constructor).
  * Consequently, `max-params` ESlint rule and `ParameterCount` Codenarc rule were updated.
* Updates README file.

## 3.0.0 - August 2019

* Breaking changes:
  * Removes `.stylelintrc.json`. (It was not providing any real benefit, with only 1 rule, that conflict with the new convention of defining quotes style locally. Besides stylelint inheritance mechanism does not lend to an easy configuration).
  * Renames `back` folder to `java`, it gives more accurate context.
  * Renames `front` folder to `js`, it gives more accurate context.
* Fixes compatibility with Checkstyle 8.21:
  * Removes `PARAMETER_DEF` from Rule `AnnotationLocationVariableAndParameter`.
  * Renames `AnnotationLocationVariableAndParameter` to `AnnotationLocationVariable`.
* Updates _Method Size Rule_ for Checkstyle, CodeNarc & ESLint to 30 lines, which is readable without scrolling, this will increase Mantainability.
* Disables `CompileStatic` Codenarc rule, since in gradle code is not mandatory.
* Adds a suppression for `e2eTest`/`E2eTest` files in `ExcessiveImports` PMD rule. (E2E tests tends to be bigger files, having only one big E2E set is useful since loading time is usually bigger and slower than other types of test.)
* Adds `.typescript-eslintrc.json`.
* Updates VERSIONS_COMPATIBILITY table.
* Updates README file.

## 2.0.7 - August 2019

* Adds `object-curly-spacing` to eslint rules.
* Improves `LineLength` Checkstyle rule to allow any number of specific tokens at start of the line.
* Improves `UseMultilineTernaryOperator` Checkstyle rule to avoid a false-positive in Method references.
* Improves `utilityClassPattern` in `ClassNamingConventions` PMD rule to also allow names defined only with one of established suffixes.
* Adds `Mappers` suffix to `utilityClassPattern` in `ClassNamingConventions` PMD rule.
* Updates `EmptyLineSeparator` Checkstyle rule to allow empty line for packages definition.
* Removes `ExcessiveMethodLength` PMD rule which is already checked by Checkstyle (which allows suppressions for tests).
* Removes `LoggerIsNotStaticFinal` PMD rule which was preventing Logger injection, useful in some special cases.

js/config/configs/eslintrc.js
//  Copyright (c) 2019 Gonzalo Müller Bravo.
//  Licensed under the MIT License (MIT), see LICENSE.txt

module.exports = {
  rules: {
    'accessor-pairs': 'error',
    'array-bracket-newline': [
      'error',
      'consistent'
    ],
    'block-scoped-var': 'error',
    'brace-style': [
      'error',
      'stroustrup'
    ],
    'camelcase': 'error',
    'complexity': [
      'error', {
        max: 15
      }
    ],
    'comma-spacing': 'error',
    'comma-style': 'error',
    'computed-property-spacing': 'error',
    'consistent-return': 'error',
    'curly': 'error',
    'dot-notation': 'error',
    'dot-location': [
      'error',
      'property'
    ],
    'eqeqeq': 'error',
    'eol-last': 'error',
    'func-call-spacing': 'error',
    'indent': [
      'error',
      2,
      {
        SwitchCase: 1,
        offsetTernaryExpressions: true,
        FunctionDeclaration: {
          body: 1,
          parameters: 2
        },
        FunctionExpression: {
          body: 1,
          parameters: 2
        },
        ignoreComments: true
      }
    ],
    'max-classes-per-file': 'error',
    'max-len': [
      'error', {
        code: 160,
        ignoreComments: true,
        ignoreTrailingComments: false,
        ignoreStrings: true,
        ignoreTemplateLiterals: true,
        ignoreRegExpLiterals: true
      }
    ],
    'max-lines': 'error',
    'max-params': [
      'error', {
        max: 7
      }
    ],
    'max-statements': [
      'error', {
        max: 30
      }
    ],
    'new-cap': [
      'error', {
        capIsNew: false
      }
    ],
    'new-parens': 'error',
    'newline-per-chained-call': 'error',
    'no-alert': 'error',
    'no-array-constructor': 'error',
    'no-caller': 'error',
    'no-duplicate-imports': 'error',
    'no-else-return': 'error',
    'no-eval': 'error',
    'no-extend-native': 'error',
    'no-extra-bind': 'error',
    'no-implicit-globals': 'error',
    'no-implied-eval': 'error',
    'no-labels': 'error',
    'no-lone-blocks': 'error',
    'no-mixed-operators': 'error',
    'no-mixed-requires': 'error',
    'no-multi-spaces': 'error',
    'no-multi-str': 'error',
    'no-new-func': 'error',
    'no-new-object': 'error',
    'no-new-require': 'error',
    'no-new-wrappers': 'error',
    'no-param-reassign': 'error',
    'no-script-url': 'error',
    'no-shadow-restricted-names': 'error',
    'no-tabs': 'error',
    'no-throw-literal': 'error',
    'no-trailing-spaces': 'error',
    'no-undef-init': 'error',
    'no-unmodified-loop-condition': 'error',
    'no-unneeded-ternary': 'error',
    'no-unused-expressions': [
      'error', {
        allowShortCircuit: true,
        allowTernary: true
      }
    ],
    'no-useless-return': 'error',
    'no-use-before-define': 'error',
    'no-void': 'error',
    'no-with': 'error',
    'object-curly-newline': [
      'error', {
        consistent: true
      }
    ],
    'object-property-newline': [
      'error',
      {
        allowAllPropertiesOnSameLine: true
      }
    ],
    'one-var': ['error', 'never'],
    'operator-assignment': 'error',
    'operator-linebreak': [
      'error',
      'before',
      {
        overrides:
        {
          '&&': 'after'
        }
      }
    ],
    'spaced-comment': [
      'error',
      'always',
      {
        exceptions: [ '-', '+', '*', '/', ':', '::', '@', '?', '?:' ],
        markers: [ '-', '+', '*', '/', ':', '::', '@', '?', '?:' ]
      }
    ],
    'space-before-function-paren': [
      'error', {
        named: 'never',
        anonymous: 'always'
      }
    ],
    'space-before-blocks': 'error',
    'space-infix-ops': 'error',
    'space-in-parens': 'error',
    'space-unary-ops': [
      'error', {
        words: true,
        nonwords: false
      }
    ],
    'wrap-iife': 'error'
  }
}
